### PR TITLE
 All multi extractors are working and can be called with functions.

### PR DESCRIPTION
**Pull request by colehurwitz**
I created two new notebooks to show how to call and use these multiextractors and added some unit tests for them.

se.concatenate_recordings_by_time(recordings, epochs) --> MultiRecordingTimeExtractor
se.concatenate_recordings_by_channel(recordings, groups) --> MultiRecordingChannelExtractor
se.concatenate_sortings(sortings) --> MultiSortingExtractor

One thing I noticed, both the MultiSorting and MultiRecordingChannel extractor change the underlying channel properties of the recordings that they concatenated... that is fine, but when you pass in duplicate recordings or sortings, the set_property calls will overwrite each other... I don't think this is a problem though because we should be expecting different extractors to be passed in.

Let me know if this pull request is ready please! @samuelgarcia @alejoe91 @magland

### CODE REVIEW

File: spikeextractors/MultiRecordingChannelExtractor.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+# Concatenates the given recordings by channel
+
+class MultiRecordingChannelExtractor(RecordingExtractor):
+    def __init__(self, recordings, groups=None):
+        RecordingExtractor.__init__(self)
+        self._RXs = recordings
+        self._all_channel_ids = []
+        self._channel_map = {}
+
+        # Sampling frequency based off the initial extractor
```
Comment (samuelgarcia) [inline spikeextractors/MultiRecordingChannelExtractor.py]: make we could check that sampling rate are all the same here ?

File: spikeextractors/MultiRecordingChannelExtractor.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        self._first_recording = recordings[0]
+        self._sampling_frequency = self._first_recording.get_sampling_frequency()
+        self._num_frames =  self._first_recording.get_num_frames()
+
+        # Test if all recording extractors have same sampling frequency
+        for i, recording in enumerate(recordings[1:]):
+            sampling_frequency = recording.get_sampling_frequency()
+            if (self._sampling_frequency != sampling_frequency):
+                raise ValueError("Inconsistent sampling frequency between extractor 0 and extractor " + str(i + 1))
+
```
Comment (samuelgarcia) [inline spikeextractors/MultiRecordingChannelExtractor.py]: Oups sorry this is done here.

File: spikeextractors/MultiRecordingChannelExtractor.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
+from .RecordingExtractor import RecordingExtractor
+import numpy as np
+
+# Concatenates the given recordings by channel
+
+class MultiRecordingChannelExtractor(RecordingExtractor):
+    def __init__(self, recordings, groups=None):
+        RecordingExtractor.__init__(self)
+        self._RXs = recordings
```
Comment (samuelgarcia) [inline spikeextractors/MultiRecordingChannelExtractor.py]: I would pu self._recordings intstead of self._RXs

File: spikeextractors/MultiRecordingChannelExtractor.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            if (self._sampling_frequency != sampling_frequency):
+                raise ValueError("Inconsistent sampling frequency between extractor 0 and extractor " + str(i + 1))
+
+        #set channel map for new channel ids to old channel ids
+        new_channel_id  = 0
+        for r_i, RX in enumerate(self._RXs):
+            channel_ids = RX.get_channel_ids()
+            for channel_id in channel_ids:
+                self._all_channel_ids.append(new_channel_id)
+                self._channel_map[new_channel_id] = {'rx': r_i, 'channel': channel_id}
```
Comment (samuelgarcia) [inline spikeextractors/MultiRecordingChannelExtractor.py]: Same here 'rx' > 'recording'
'channel' > 'channel_id'

File: spikeextractors/MultiRecordingChannelExtractor.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                new_channel_id += 1
```
Comment (alejoe91) [inline spikeextractors/MultiRecordingChannelExtractor.py]: Maybe here you could check if the channels have duplicates. For instance, if you concatenated 2 recordings with channels: [0,1,2,3], [8,9,10,11] the channel ids should be maintained

File: spikeextractors/extractors/tridescloussortingextractor/tridescloussortingextractor.py [hunk 6]
```diff
@@ -47,4 +47,4 @@ def get_unit_spike_train(self, unit_id, start_frame=None, end_frame=None):
             spike_times = spike_times[spike_times >= start_frame]
         if end_frame is not None:
             spike_times = spike_times[spike_times < end_frame]
-        return spike_times.copy()  #�copy avoid reference to the unerlying memmap
+        return spike_times.copy()  #�copy avoid reference to the unerlying memmap
```
Comment (alejoe91) [inline spikeextractors/extractors/tridescloussortingextractor/tridescloussortingextractor.py]: ?

**Comment by alejoe91** [comment]
Looks good @colehurwitz 

Just had some comments :)

**Comment by samuelgarcia** [comment]
Hi @colehurwitz .
I am not evry familiar yet with MultiRecordingExtractor.
Your PR looks good. 
I made few comments.

Globaly, I don't like the general approach of this MultiRecordingExtractor (time and channel).
This comment is outside this PR.
Because it concatenate in memory evrything. This is won't work for big dataset.
If we need to concatenate, 3 recording of 1 hour the among of memory needed to make this is phenomenal for high channel count.
I think I prefer the explicit approach of multi segment neo like where the user need to loop explicitly on recorings (segment)

**Comment by colehurwitz** [comment]
Hi all! Thanks for the suggestions, I will correct a few things about my PR.

Unless I am mistaken @samuelgarcia, we only ever load things into memory when calling get_traces and even then, we would never load the whole recording in. The MulitRecording extractors just coordinate extraction from multiple files.

**Comment by samuelgarcia** [comment]
@colehurwitz : yes it is true.
I don't known how to explain but I prefer explicit.
This magic auto concatenate do not make happy I prefer API where looping over recording sis explicit and not hidden.
phy/klust/sc have the same approach of magic concatenate and I don't like it so much.
End user of spikeextractors are devs not experimentalist, so explicit loop over segment is no so hard.

Nevertheless, I already pronounce myself on this and this is not related to this PR.
So no worry.

**Comment by alejoe91** [comment]
I agree with Cole. Only when get_traces is called (for the required start_frame and end_frame) the data is actually loaded in memory.

If you want to concatenate to spikesort, you have to dump all the files in the same output file, at least for most sorters

**Comment by samuelgarcia** [comment]
> If you want to concatenate to spikesort, you have to dump all the files in the same output file, at least for most sorters
Because the only sorter that deal with multi segment is tdc :)
All others play with a one segment only.

We could push a layer that would help devs to loop explicitly over segments.
here the magic concatenate is a trap because  at the end of the pipe spiek index are absolut and we need a trick to assign them back to the related file and segment which is for the end and (experimentalist) a pain.

As a end user, to retrieve spike I would prefer a list of sorting that explicitly is the same size of the recording list and not spike indexe that relate to the concatenated recordings.

**Comment by colehurwitz** [comment]
I see your point @samuelgarcia. I think why we liked the auto-magic approach is that spike sorters and tools can just treat the Multi-extractors the same way as normal extractors which makes development a little easier in my opinion :) 

However, the explicit approach might help us avoid issues in the future that I am not yet aware of. It is something to keep in mind for sure!

**Comment by samuelgarcia** [comment]
>  to keep in mind for sure!

Trust i have this in mind for the futur :)

**Comment by colehurwitz** [comment]
We could maybe add a function to the Multi extractor like: ‘get_traces_explicit’ which makes the call an explicit one or something! Something I will think about as well

**Comment by alejoe91** [comment]
I would say let's merge for now and we can implement the `segment` approach later :)

**Comment by samuelgarcia** [comment]
Of course.
Maybe next year. We more urgent thing to do now (5 papers!!)

File: spikeextractors/MultiRecordingChannelExtractor.py (hunk 5 above)
Comment (colehurwitz) [reply spikeextractors/MultiRecordingChannelExtractor.py]: I would rather remap I think just to keep the behaviour consistent

File: spikeextractors/extractors/tridescloussortingextractor/tridescloussortingextractor.py (hunk 6 above)
Comment (colehurwitz) [reply spikeextractors/extractors/tridescloussortingextractor/tridescloussortingextractor.py]: I'll fix this in the next pull request I make. Weird!

**Pull request merged**